fieldpath/fromvalue.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldpath
+
+import (
+	"github.com/kubernetes-sigs/structured-merge-diff/value"
+)
+
+// SetFromValue creates a set including each field that's set in the value.
+func SetFromValue(v value.Value) *Set {
+	s := NewSet()
+
+	w := objectWalker{
+		path:  Path{},
+		value: v,
+		do:    func(p Path) { s.Insert(p) },
+	}
+
+	w.walk()
+	return s
+}
+
+type objectWalker struct {
+	path  Path
+	value value.Value
+
+	do func(Path)
+}
+
+func (w *objectWalker) walk() {
+	if len(w.path) > 0 {
+		w.do(w.path)
+	}
+
+	switch {
+	// Nothing to do for any of these:
+	case w.value.Null:
+	case w.value.Float != nil:
+	case w.value.Int != nil:
+	case w.value.String != nil:
+	case w.value.Boolean != nil:
+
+	// Descend
+	case w.value.List != nil:
+		for i, child := range w.value.List.Items {
+			w2 := *w
+			w2.path = append(w.path, GuessBestListPathElement(i, child))
+			w2.value = child
+			w2.walk()
+		}
+	case w.value.Map != nil:
+		for _, child := range w.value.Map.Items {
+			w2 := *w
+			w2.path = append(w.path, PathElement{FieldName: &child.Name})
+			w2.value = child.Value
+			w2.walk()
+		}
+	}
+}
+
+// AssociativeListCandidateFieldNames lists the field names which are
+// considered keys if found in a list element.
+var AssociativeListCandidateFieldNames = []string{
+	"key",
+	"id",
+	"name",
+}
+
+// GuessBestListPathElement guesses whether item is an associative list
```

Comment:
In an ideal world, we would have great tool and people would do the right thing, and we wouldn't have to guess anything, because we would have all the information we need to do what must be done.

But we're not in an ideal world, is this a good intermediary state that brings us closer to one? I'm not completely convinced. Guessing usually involves lots of "bugs" and confusion that I'd rather avoid if necessary.

Reply:
The important thing is, people can fix the bugs by providing a schema.

Most CRDs are currently schemaless; apply needs to do something, it can't fail completely for them.

Reply:
> The important thing is, people can fix the bugs by providing a schema.